### ISSUE TITLE
[query/bug] mark testcase as failed and stop execution when `t.FailNow()` is encountered in an assessment

### ISSUE THREAD

**Issue by divmadan** [kubernetes-sigs/e2e-framework#112]
[Release v0.0.5]

I have a requirement to stop the testcase execution if certain assertions during the course of the testcase fails. I thought that we could control this behaviour by using [t.FailNow](https://pkg.go.dev/testing#T.FailNow) but it just marks that assertion as `FAIL` (thus behaving as [t.Fail](https://pkg.go.dev/testing#T.Fail)) and continues execution at the next assertion/feature.

Looks like it's due to the local scope of the variable `t *testing.T` in the assertion definition. 

Is this the desired behaviour? How should I work out my requirement, workarounds or suggestions?

***

MWE:

```go
// example_test.go

package example

import (
	"context"
	"testing"

	"sigs.k8s.io/e2e-framework/pkg/envconf"
	"sigs.k8s.io/e2e-framework/pkg/features"
)

func TestExample(t *testing.T) {
	failFeature := features.New("fail-feature").
		Assess("1==2", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
			if 1 != 2 {
				t.Log("1 != 2")
				t.FailNow() // mark test case as failed here, don't continue execution
			} else {
				t.Log("1 == 2")
			}
			return ctx
		}).
		Assess("print", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
			t.Log("THIS LINE SHOULDN'T BE PRINTED")
			return ctx
		}).
		Feature()

	nextFeature := features.New("next-feature").
		Assess("print", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
			t.Log("THIS LINE ALSO SHOULDN'T BE PRINTED")
			return ctx
		}).
		Feature()

	testenv.Test(t, failFeature, nextFeature)
}
```

```go
// main_test.go

package example

import (
	"log"
	"os"
	"path/filepath"
	"testing"

	"sigs.k8s.io/e2e-framework/pkg/env"
)

var testenv env.Environment

func TestMain(m *testing.M) {
	// get kubeconfig
	home, err := os.UserHomeDir()
	if err != nil {
		log.Fatal("error while getting user home directory")
	}
	testenv = env.NewWithKubeConfig(filepath.Join(home, ".kube", "config"))

	os.Exit(testenv.Run(m))
}
```

```
➜ go test . -v
=== RUN   TestExample
=== RUN   TestExample/fail-feature
=== RUN   TestExample/fail-feature/1==2
    example_test.go:15: 1 != 2
=== RUN   TestExample/fail-feature/print
    example_test.go:23: THIS LINE SHOULDN'T BE PRINTED
=== RUN   TestExample/next-feature
=== RUN   TestExample/next-feature/print
    example_test.go:30: THIS LINE ALSO SHOULDN'T BE PRINTED
--- FAIL: TestExample (0.00s)
    --- FAIL: TestExample/fail-feature (0.00s)
        --- FAIL: TestExample/fail-feature/1==2 (0.00s)
        --- PASS: TestExample/fail-feature/print (0.00s)
    --- PASS: TestExample/next-feature (0.00s)
        --- PASS: TestExample/next-feature/print (0.00s)
FAIL
FAIL    example.com     0.532s
FAIL
```

**Comment by harshanarayana** [comment]
Did you try using `-failfast` arg with the `go test` ?

```bash
❯ go test . -v -failfast
=== RUN   TestExample
=== RUN   TestExample/fail-feature
=== RUN   TestExample/fail-feature/1==2
    fail_test.go:15: 1 != 2
--- FAIL: TestExample (0.00s)
    --- FAIL: TestExample/fail-feature (0.00s)
        --- FAIL: TestExample/fail-feature/1==2 (0.00s)
FAIL
FAIL	sigs.k8s.io/e2e-framework/examples/failfast	1.601s
FAIL

```

**Comment by vladimirvivien** [comment]
Thanks for pointing this out. There is definitely a need to revisit the way the test context `testing.T` is managed internally with respect to features.  Since features are like sub-tests, the parent test context (from the test function) is not affected, causing features test to not fail when expected.

For now, try the workaround from @harshanarayana mentioned.  We will investigate a better approach.

**Comment by divmadan** [comment]
@harshanarayana @vladimirvivien 

Thanks for the prompt response.

`failfast` leads to exit of the whole suite. I still want to continue executing the other testcases, only I want the failing testcase to exit.

Adding to the above example, 

```go
// example_test.go

package example

import (
	"context"
	"testing"

	"sigs.k8s.io/e2e-framework/pkg/envconf"
	"sigs.k8s.io/e2e-framework/pkg/features"
)

func TestExample(t *testing.T) {
	failFeature := features.New("fail-feature").
		Assess("1==2", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
			if 1 != 2 {
				t.Log("1 != 2")
				t.FailNow() // mark test case as failed here, don't continue execution
			} else {
				t.Log("1 == 2")
			}
			return ctx
		}).
		Assess("print", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
			t.Log("THIS LINE SHOULDN'T BE PRINTED")
			return ctx
		}).
		Feature()

	nextFeature := features.New("next-feature").
		Assess("print", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
			t.Log("THIS LINE ALSO SHOULDN'T BE PRINTED")
			return ctx
		}).
		Feature()

	testenv.Test(t, failFeature, nextFeature)
}

// even if the previous testcase fails, execute this testcase
func TestNext(t *testing.T) {
	nextFeature := features.New("next-test-feature").
		Assess("print", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
			t.Log("THIS LINE SHOULD BE PRINTED")
			return ctx
		}).
		Feature()

	testenv.Test(t, nextFeature)
}
```

Here, when `TestExample/fail-feature` fails, I want the testcase `TestExample` to fail and stop execution only for it. I still want to execute the upcoming `TestNext` testcase.

Current behaviour:
```
➜ go test . -v -failfast
=== RUN   TestExample
=== RUN   TestExample/fail-feature
=== RUN   TestExample/fail-feature/1==2
    example_test.go:15: 1 != 2
--- FAIL: TestExample (0.00s)
    --- FAIL: TestExample/fail-feature (0.00s)
        --- FAIL: TestExample/fail-feature/1==2 (0.00s)
FAIL
FAIL    example.com     0.292s
FAIL
```

**Comment by vladimirvivien** [comment]
@divmadan this is great. Thank you for the great explanation and I think you are correct. The failing assessment should stop without taking down the entire test suite or even the entire TestFunction where it is embedded.

I am going to mark this as an important bug so that it gets the attention it deserves.

**Comment by divmadan** [comment]
@vladimirvivien 

> The failing assessment should stop without taking down the entire test suite or even the entire TestFunction where it is embedded.

Agreed.

***

> or even the entire TestFunction where it is embedded.

In my case, there's still a requirement to exit the test function as the features are dependent on their predecessors.

As you mentioned that the features are like sub-tests, I'll assume that you are leaning towards failing a feature when a `t.FailNow()` is encountered within an assertion. In that case, I'll still need another way to control the behaviour of the entire testcase. Please do point out if the assumption is incorrect.

PS: If there's a way to chain features to signify dependencies between them, that would be great and it'll eliminate the need to control the behaviour of the testcase.

**Comment by harshanarayana** [comment]
@vladimirvivien  @divmadan 

With a few minor changes to the framework code, I was able to get the following behavior as intended. 

```bash
=== RUN   TestExample
=== RUN   TestExample/fail-feature
=== RUN   TestExample/fail-feature/1==2
    example_test.go:15: 1 != 2
--- FAIL: TestExample (0.00s)
    --- FAIL: TestExample/fail-feature (0.00s)
        --- FAIL: TestExample/fail-feature/1==2 (0.00s)



=== RUN   TestNext
=== RUN   TestNext/next-test-feature
=== RUN   TestNext/next-test-feature/print
    example_test.go:42: THIS LINE SHOULD BE PRINTED
--- PASS: TestNext (0.00s)
    --- PASS: TestNext/next-test-feature (0.00s)
        --- PASS: TestNext/next-test-feature/print (0.00s)
FAIL
```

However, this brings up a few questions about how do we go around to getting this in

1. Can we expose a `t.FailNow` like construct from `feature` that is passed around to the assessment that can be used to explicitly fail the feature as a whole ?
2.  Do we expose a global flag that controls this behavior and based on `t.Failed()` we take the necessary decision ? (this is how I got around to getting it working for now)
3. What granularity do we want to provide a fail handlers at ?
4. Currently, we have `Test` and `TestInParallel` handlers. We can also provide a `TestWithRequiredDep` for example that will ensure the previous feature is tested with `t.Failed() != true` and only then will it continue. But this gets a bit messy with `--parallel` flag enabled.
